wallet-new/src/Cardano/Wallet/Types/UtxoStatistics.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE DeriveGeneric #-}
+
+module Cardano.Wallet.Types.UtxoStatistics
+    ( computeUtxoStatistics
+    , UtxoStatistics (..)
+    , HistogramBar (..)
+    , BoundType (..)
+    , generateBounds
+    )  where
+
+
+import           Universum
+
+import qualified Control.Foldl as L
+import           Control.Lens (at, (?~))
+import           Data.Aeson
+import qualified Data.HashMap.Strict as HMS
+import qualified Data.List.NonEmpty as NL
+import qualified Data.Map.Strict as Map
+import           Data.Scientific (floatingOrInteger)
+import           Data.Swagger hiding (Example, example)
+import qualified Data.Text as T
+import           Data.Word (Word64)
+import           Formatting (bprint, build, (%))
+import qualified Formatting.Buildable
+import           Serokell.Util (listJson)
+import           Test.QuickCheck
+
+import           Cardano.Wallet.API.V1.Swagger.Example (Example)
+import           Pos.Infra.Util.LogSafe (BuildableSafeGen (..),
+                     deriveSafeBuildable)
+
+
+--  Utxo statistics for the wallet.
+--  Histogram is composed of bars that represent the bucket. The bucket is tagged by upper bound of a given bucket.
+--  The bar value corresponds to the number of stakes
+--  In the future the bar value could be different things:
+--  (a) sum of stakes in a bucket
+--  (b) avg or std of stake in a bucket
+--  (c) topN buckets
+--  to name a few
+data HistogramBar = HistogramBarCount
+    { bucketName       :: !Text
+    , bucketUpperBound :: !Word64
+    } deriving (Show, Eq, Ord, Generic)
+
+--  Buckets boundaries can be constructed in different way
+data BoundType = Log10 | Haphazard
+
+generateBounds :: BoundType -> NonEmpty Word64
+generateBounds bType =
+    let (^!) :: Word64 -> Word64 -> Word64
+        (^!) = (^)
+    in case bType of
+        Log10 -> NL.fromList $ ( map (\toPower -> 10 ^! toPower) [1..16] ) ++ [45 * (10 ^! 15)]
+        Haphazard -> NL.fromList [10, 100, 1000, 10000]
+
+instance Arbitrary HistogramBar where
+    arbitrary =
+        let possibleBuckets = fmap show (generateBounds Log10)
+            possibleBars = NL.zipWith HistogramBarCount possibleBuckets (NL.fromList [0..])
+        in elements (NL.toList possibleBars)
```

Comment:
I don't think this does what you want -- this makes it choose a bucket name and a constant number from the `[0..]` list. Eg, for every `HistogramBarCount x y`, the `x` will uniquely determine the `y`.

I think what you want instead is something like:

```haskell
arbitrary = do
    name <- elements $ map show (generateBounds Log10)
    bound <- arbitrary
    pure (HistogramBarCount name bound)
```

This will pick a random upper bound and name from the list.

Reply:
done